### PR TITLE
feat: initial keyboard shortcut support

### PR DESCRIPTION
**Pull request by gkatsev**
Kicking off the work for https://github.com/muxinc/media-chrome/discussions/257

### CODE REVIEW

File: src/js/media-container.js
Comment (gkatsev) [inline src/js/media-container.js:414]: all of these were copied from the corresponding buttons. It would be nice if we can share the code, but I'm not positive there's a great way to do it. We could do something like `PlayButton.prototype.handleClick.call(this, e)` or something. Thoughts?

File: src/js/media-container.js
Comment (cjpillsbury) [inline src/js/media-container.js:354]: suggestion(blocker): We should use the keydown + keyup strategy here or similar to filter out false positives while meta keys are pressed https://github.com/muxinc/media-chrome/blob/main/src/js/media-chrome-button.js#L108-L126 . If/when we add custom mappings, we'd still likely want "exact matches" even for multi-key support

Comment (cjpillsbury) [inline src/js/media-container.js:354]: suggestion(blocker): Before merging/releasing, let's make sure there's some way for implementors to disable this functionality. Can be a simple attribute like `disableshortcuts` (or whatever name is appropriate). Should also be an `observedAttribute` to ensure that the relevant event handler(s) are added/removed.

**Review by cjpillsbury — CHANGES_REQUESTED** [review]
a couple of asks, but this looks close to done for a good mvp spike.

File: src/js/media-container.js
Comment (cjpillsbury) [reply src/js/media-container.js:414]: Glad you're thinking of that. For now I'd lean towards "duplicate code" for this PR and KISS functional-like approaches (some module that "owns" a bucket of functions that take an element and return an appropriate `CustomEvent` (imo the "`dispatchEvent`" side effect should be externalized, but I could be convinced otherwise)

File: src/js/media-container.js
Comment (gkatsev) [reply src/js/media-container.js:409]: yeah, this makes sense. I think also worth calling it out in https://github.com/muxinc/media-chrome/discussions/257.
This actually brings up an interesting point in that the code is currently in `media-container` and not `media-controller` and an attribute and default value should probably live in `media-controller`?

I guess one way of handling that is that we call the `this.handleKeyupEvent` or whatever method on `this` but only implement it in `media-controller` with an no-op method in `media-container`.

Comment (cjpillsbury) [reply src/js/media-container.js:409]: yeah that's a good point. These interactions are more about "state management"/"state change requests," so they probably belong in `media-controller`. I also think it would be fine to add a second event handler there for keyup, fwiw (which may make more sense anyway if you do key down + key up pairs similar to the components).

File: src/js/media-container.js
Comment (heff) [inline src/js/media-container.js:20]: This makes me cringe a little. Partly for the added dependency complexity, but partly because it means we've made media chrome's button and range more special somehow in this ecosystem. I'm not gonna call this a blocker but want to encourage us to consider an approach that removes this dependency, maybe as a fast follow.

Comment (gkatsev) [reply src/js/media-container.js:20]: Yeah, this is mostly for to get it working. I think the proper solution is what we discussed in https://github.com/muxinc/media-chrome/discussions/257, where the elements can expose a property like `keysUsed` and then below, we can do
```js
// skip handler is already used by the target
if (evt.target.keysUsed ?? []).includes(evt.key)) {
  return;
}
```
And we won't need to import these directly.

The alternative is that each component that handles keyup will call `stopPropagation` properly.

Originally, I was thinking we can punt on this, but this affects a component that's currently in Mux Player, so, we probably should come to a decision sooner rather than earlier. I lean towards a properly like `keysUsed` (or another name).

File: src/js/media-container.js
Comment (gkatsev) [reply src/js/media-container.js:354]: There doesn't seem to be a very good precedent for this, but I would lean towards `nokeyboardshortcuts` or `noshortcuts`, potentially with `globalkeyboardhosrtcuts` or `globalshortcuts` for later attaching to `window`.

File: src/js/media-container.js
Comment (heff) [reply src/js/media-container.js:20]: Ok cool, makes sense. 👍 

Without understanding all the complexity, I still feel better about something where control-specific complexity is kept in the controls, or the super classes of controls, rather than making media container/controller smarter here so controls don't have to be. Probably worth a call to discuss.

File: src/js/media-controller.js
Comment (gkatsev) [inline src/js/media-controller.js:499]: should this go into `constants.js` somewhere? MediaUIAttributes doesn't seem correct.
Also, should it be `media-noshortcuts`?

File: src/js/media-controller.js
Comment (gkatsev) [inline src/js/media-controller.js:57]: moved from the bottom of the class here, which seems more appropriate.

File: src/js/media-controller.js
Comment (gkatsev) [inline src/js/media-controller.js:21]: The `keysUsed` vs `stopPropagation` should be resolved before this PR can be merged.

File: src/js/media-container.js
Comment (gkatsev) [reply src/js/media-container.js]: added a note here https://github.com/muxinc/media-chrome/discussions/257#discussioncomment-3132937

File: src/js/media-container.js
Comment (gkatsev) [reply src/js/media-container.js]: https://github.com/muxinc/media-chrome/pull/263#discussion_r919492637

File: src/js/media-container.js
Comment (gkatsev) [reply src/js/media-container.js]: Added a new comment so that this doesn't get lost here https://github.com/muxinc/media-chrome/pull/263#discussion_r919493069

File: src/js/media-controller.js
Comment (gkatsev) [reply src/js/media-controller.js:21]: I created two PRs with examples of `keysUsed` and `stopPropagation`:
- keysUsed https://github.com/gkatsev/media-chrome/pull/1
- stopPropagation https://github.com/gkatsev/media-chrome/pull/2

To me, keysUsed is actually a bit less presumptuous since it requires less additional code from components, and it doesn't prevent users from still being able to listen to the key events if they should want.
In comparison, by default, with the stopPropagation thing, we either need to have special handling for native elements like buttons and inputs with hardcoded keys that are used, or our wrapper needs to specifically call stopPropagation, otherwise, both the native behavior and the keyboard shortcut would end up triggering. And now, we're also limiting the ability of users to be able to listen to keyup with these as targets.

Comment (gkatsev) [reply src/js/media-controller.js:21]: follow-up from chat:
- write some docs on how this would be used by folks
- should this be an attribute instead of a property? Or should it be both?
- better name than keysUsed?
- keysUsed doesn't preclude from media-chrome users using stopPropagation in their keyboard handling to prevent multiple behaviors from occurring.

File: src/js/media-controller.js
Comment (luwes) [reply src/js/media-controller.js:499]: I like the compactness of `noshortcuts`, but not super clear to me right off the bat. why not `nokeyboard`?

File: src/js/media-controller.js
Comment (luwes) [reply src/js/media-controller.js:21]: I'm leaning more towards stopPropagation, seems this is exactly meant for this and no extra API is added. less coupling. the little extra code in the components is worth that I think.

File: src/js/media-controller.js
Comment (gkatsev) [reply src/js/media-controller.js:499]: Alternative name: `nohotkeys`

File: src/js/media-controller.js
Comment (gkatsev) [reply src/js/media-controller.js:21]: Agreed on both property `keysUsed` and attribute `keyused`.

File: src/js/media-controller.js
Comment (cjpillsbury) [inline src/js/media-controller.js:743]: question(non-blocking): Do we need these checks?

File: src/js/media-controller.js
Comment (cjpillsbury) [inline src/js/media-controller.js:748]: thought(non-blocking): As a followup, we'll need to figure out how this should work for non-nested components based on the current architecture (See code/architecture for `associatedElements` and how event (re)propagation works).

File: src/js/media-controller.js
Comment (cjpillsbury) [inline src/js/media-controller.js:496]: suggestion(non-blocking): Instead of continuing to listen to irrelevant keyboard presses when `"nohotkeys"` is set, we should probably:

1. add `"nohotkeys"` to `observedAttributes`.
2. add/remove event handlers for `"keyup"`/`"keydown"` based on `attributeChangedCallback`
    1. This entails moving the event callbacks to methods instead of inline.

File: docs/media-controller.md
Comment (cjpillsbury) [inline docs/media-controller.md:311]: chore(blocking):

1. Discuss and link to https://developer.mozilla.org/en-US/docs/Web/API/UI_Events/Keyboard_event_key_values
2. Callout the `"Space"` keyword mapping

**Review by cjpillsbury — CHANGES_REQUESTED** [review]
One docs chore and a handful of non-blocking callouts. Otherwise, 👌. Did basic smoke testing across available browsers.

File: src/js/media-controller.js
Comment (gkatsev) [reply src/js/media-controller.js:743]: Maybe not, since I guess the event bubbling should guarantee that it's coming from a child element?

File: src/js/media-controller.js
Comment (gkatsev) [reply src/js/media-controller.js:748]: yup, this got called out in the discussion #257

File: src/js/media-controller.js
Comment (gkatsev) [reply src/js/media-controller.js]: updated

File: docs/media-controller.md
Comment (gkatsev) [reply docs/media-controller.md]: updated

File: docs/media-controller.md
Comment (cjpillsbury) [inline docs/media-controller.md:314]: 😎

File: src/js/media-controller.js
Comment (cjpillsbury) [reply src/js/media-controller.js:767]: yup that's what I was thinking. We may also get into some weirdness with `contains` checks for nested web component use cases.

File: src/js/media-controller.js
Comment (cjpillsbury) [inline src/js/media-controller.js:741]: issue(blocker): missed `this.#keyUpHandler` refactor.

**Review by cjpillsbury — APPROVED** [review]
pre-approving with the minor callout and recommend to remove the contains et al. check.

File: docs/media-controller.md
Comment (luwes) [inline docs/media-controller.md:90]: 👍

**Review by luwes — APPROVED** [review]
LGTM! 🎉

**Pull request merged**
